router/fastdp.go

```diff
@@ -1171,6 +1179,45 @@ func (fastdp *FastDatapath) takeDecoder(lock *fastDatapathLock) *EthernetDecoder
 	return dec
 }
 
+// A checkFlowLoop checks whether the flow is free from the following loops:
+// * in_port == out_port, where in_port is non-vxlan vport;
+// * a packet is sent back to a vxlan tunnel it has been received from.
+func (fastdp *FastDatapath) checkFlowLoop(flow *odp.FlowSpec) bool {
+	var vxlanKey odp.TunnelAttrs
+
+	inVports := make(map[odp.VportID]struct{})
+	inVxlan := false
+
+	for _, key := range flow.FlowKeys {
+		switch k := key.(type) {
+		case odp.InPortFlowKey:
+			inVports[k.VportID()] = struct{}{}
+		case odp.TunnelFlowKey:
+			inVxlan = true
+			vxlanKey = k.Key()
+		}
+	}
+
+	for _, action := range flow.Actions {
+		switch a := action.(type) {
+		case odp.OutputAction:
+			if _, ok := inVports[a.VportID()]; ok {
+				if _, ok := fastdp.vxlanVportIDs[a.VportID()]; !ok {
+					return false
+				}
+			}
+		case odp.SetTunnelAction:
+			if inVxlan && a.TunnelAttrs.TunnelId == vxlanKey.TunnelId &&
```

Comment:
I was referring to the case when `relay`ing a packet, i.e. the packet's final dst is known, so the VNI does not change.

However, there is a second case, which I did not consider, when we do `relayBroadcast`, i.e. the dst peer changes, so the VNI does as well.

I suggest to modify the condition by replacing it with `(a.TunnelAttrs.TunnelId == vxlanKey.TunnelId || a.TunnelAttrs.TunnelId == reverse(vxlanKey.TunnelId)`.